### ISSUE TITLE
Service Validator does not flag links to non-subordinate resource outside 'Links' object.

### ISSUE THREAD

**Issue by pointbazaar** [DMTF/Redfish-Service-Validator#663]
Hi i have a question / issue w.r.t. reference properties

```
3.3.16 subordinate resource
is subordinate to another resource if the initial segment of the resource URI is the same as the URI of the other
resource, but is at least one level deeper
Note 1 to entry: For example, /redfish/v1/Chassis/A88/Assembly is a subordinate resource of the Chassis resource
named A88 .

...

9.5.7 Reference properties
Reference properties provide a reference to another resource in the data model. Reference properties are JSON
objects that contain an @odata.id property. The @odata.id property value is the URI of the referenced resource.

...

9.6.13 Links
The Links property represents the hyperlinks associated with the resource, as defined by that resource's schema
definition. All associated reference properties defined for a resource shall be nested under the links property. All
directly (subordinate) referenced properties defined for a resource shall be in the root of the resource. There are
some exceptions to these rules, such as where ease of expansion or deep operations is beneficial to the user.
```

In my example i put the following in a response for a ComputerSystem resource GET

```cpp
res.jsonValue["LogServices"]["@odata.id"] = "/redfish/v1/Managers/bmc/LogServices"
```

```bash
curl --insecure --user root:**** https://${bmc}/redfish/v1/Systems/Tyan_S8030_Baseboard
```
```json
{
  "@odata.id": "/redfish/v1/Systems/Tyan_S8030_Baseboard",
  "@odata.type": "#ComputerSystem.v1_22_0.ComputerSystem",
  "Actions": {
    "#ComputerSystem.Reset": {
      "@Redfish.ActionInfo": "/redfish/v1/Systems/Tyan_S8030_Baseboard/ResetActionInfo",
      "target": "/redfish/v1/Systems/Tyan_S8030_Baseboard/Actions/ComputerSystem.Reset"
    }
  },
  "Boot": {
    "AutomaticRetryConfig": "RetryAttempts",
    "AutomaticRetryConfig@Redfish.AllowableValues": [
      "Disabled",
      "RetryAttempts"
    ],
    "BootSourceOverrideEnabled": "Disabled",
    "BootSourceOverrideMode": "UEFI",
    "BootSourceOverrideMode@Redfish.AllowableValues": [
      "Legacy",
      "UEFI"
    ],
    "BootSourceOverrideTarget": "None",
    "BootSourceOverrideTarget@Redfish.AllowableValues": [
      "None",
      "Pxe",
      "Hdd",
      "Cd",
      "Diags",
      "BiosSetup",
      "Usb"
    ],
    "StopBootOnFault": "Never",
    "TrustedModuleRequiredToBoot": "Disabled"
  },
  "Description": "Computer System",
  "GraphicalConsole": {
    "ConnectTypesSupported": [
      "KVMIP"
    ],
    "MaxConcurrentSessions": 4,
    "ServiceEnabled": true
  },
  "HostWatchdogTimer": {
    "FunctionEnabled": false,
    "Status": {
      "State": "Enabled"
    },
    "TimeoutAction": "PowerCycle"
  },
  "Id": "Tyan_S8030_Baseboard",
  "LastResetTime": "2026-01-26T02:23:22+00:00",
  "Links": {
    "ManagedBy": [
      {
        "@odata.id": "/redfish/v1/Managers/bmc"
      }
    ]
  },
  "LocationIndicatorActive": false,
  "LogServices": {
    "@odata.id": "/redfish/v1/Managers/bmc/LogServices"
  },
  "MemorySummary": {
    "TotalSystemMemoryGiB": 0.0
  },
  "Name": "Tyan_S8030_Baseboard",
  "PowerRestorePolicy": "AlwaysOff",
  "PowerState": "Off",
  "ProcessorSummary": {
    "Count": 0
  },
  "SerialConsole": {
    "IPMI": {
      "ServiceEnabled": true
    },
    "MaxConcurrentSessions": 15,
    "SSH": {
      "HotKeySequenceDisplay": "Press ~. to exit console",
      "Port": 2200,
      "ServiceEnabled": true
    }
  },
  "Status": {
    "Health": "OK",
    "State": "Disabled"
  },
  "SystemType": "Physical"
}
```

where `/redfish/v1/Managers/bmc/LogServices` is not a subordinate resource.

One interpretation  of `9.6.13 Links` would suggest that it is not allowed since

```
  "LogServices": {
    "@odata.id": "/redfish/v1/Managers/bmc/LogServices"
  },
```

is not under `Links` and can only reference subordinate resources.

In my experience the Service Validator did not check for this.

I am asking if my interpretation here is correct and if the validator should be patched to catch errors like that.

**Comment by mraineri** [comment]
I'll need to raise this with the forum. Technically per the schema definition this is valid, but I agree in this case this does not seem like a correct usage of the property. One thing to keep in mind is in some circumstances we do allow for redirecting certain links that are subordinate from a modeling perspective to another URI tree. Take the composability usage as an example; many collections under a system will redirect to resource blocks to show the components that comprise the system (https://github.com/DMTF/Redfish-Publications/blob/main/mockups/public-composability/Systems/ComposedSystem/Processors/index.json). So, adding testing for this specific scenario may be problematic. The language in the spec was more for DMTF and other schema writers to follow best practices when drafting new URIs to add to the model rather than

**Comment by pointbazaar** [comment]
Thanks for this quick response and the example 👍🏼 

> So, adding testing for this specific scenario may be problematic.

If there are exceptions, would it work to specify it for each reference property as an attribute in the schema?

cc @edtanous

**Comment by pointbazaar** [comment]
@mraineri  sorry to ping - did you have a chance to create a forum post - if so, can you send a link?

Otherwise should i go ahead and create a forum post regarding the validator issue and solution proposal?

I am not super focused on fixing it in the validator but more so to first have some information source to determine if it's okay for a given property (not under `Links`) to reference non-subordinate resource, as guidance for developers writing redfish implementations like bmcweb.

Perhaps something like `"canReferenceNonsubordinateResource": true` on the property?

**Comment by mraineri** [comment]
For this type of issue, we need to discuss this within DMTF first. Something like this is likely going to require spec or schema changes to convey requirements for when we need strict usage of "subordinate" URIs. However, I'm also concerned we're too late in the game for Redfish to even mandate something like this.

I'll be brining this topic up.
